File: Java_ws/ProgettoJava/src/cacciatori/Cacciatore.java
package cacciatori;

import java.util.ArrayList;
import java.util.List;
import animali.Cacciabile;

public final class Cacciatore implements Comparable<Cacciatore> {
	private static int count = 0; //campo statico per generare l'id

	//CAMPI
	final String id;
	String nome;
	int punti;
	List<Cacciabile> trofei;

	//COSTRUTTORE
	public Cacciatore(String nome) {
		this.id = "C" + (++count);
		this.nome = nome;
		this.punti = 0;
		this.trofei = new ArrayList<Cacciabile>();
	}
	
	//METODI
	public String getId() {
		return id;
	}
	public String getNome() {
		return nome;
	}
	public void setNome(String nome) {
		this.nome = nome;
	}
	public int getPunti() {
		return punti;
	}
	public int addPunti(int p) {
		return this.punti += p;
	}
	//il trofeo non può essere modificato, lo faranno altri metodi
	public void addTrofeo(final Cacciabile trofeo) {
		trofei.add(trofeo);
	}
	@Override
	public String toString() {
		return "Cacciatore " + nome + "\n   id = " + id + "\n   punti = " + punti + "\n   numero trofei = " + trofei.size();
	}

	//ordinamento di default dipende dal solo id univoco
	@Override
	public int compareTo(Cacciatore o) {
		return this.id.compareTo(o.id);
	}
}

File: ProgettoCpp/src/Cacciatore.cpp
#include "Cacciatore.h"

namespace cacciatori {

int Cacciatore::count = 0; //inizializzazione variabile statica

Cacciatore::Cacciatore(string n): id("C" + to_string(++count)), nome(n), punti(0)  { }

const string& Cacciatore::getId() const {
	return id;
}

const string& Cacciatore::getNome() const {
	return nome;
}

void Cacciatore::setNome(const string& nome) {
	this->nome = nome;
}

int Cacciatore::getPunti() const {
	return punti;
}

void Cacciatore::addPunti(int p) {
	this->punti += p;
}

void Cacciatore::addTrofeo(shared_ptr<animali::Animale> trofeo) {
	this->trofei.push_back(trofeo);
}

//overload dell'operatore <<
ostream& operator<<(ostream& out, const Cacciatore& cacciatore) {
	out << "Cacciatore " << cacciatore.nome << endl;
	out << "   id = " << cacciatore.id << endl;
	out << "   punti = " << cacciatore.punti << endl;
	out << "   numero trofei = " << cacciatore.trofei.size() << endl;
	return out;
}

} /* namespace cacciatori */
